### ISSUE TITLE
RFC: Streaming Conformer with KV Caching (`muaalem_streaming_kv_cache`)

### ISSUE THREAD

**Issue by obadx** [obadx/quran-muaalem#30]
## High-Level Goal

Create a new module parallel to `muaalem_offline` and `muaalem_streaming_lstm` that implements a streaming Conformer encoder using KV caching, supporting both full-context and sliding-window attention. Input: 250 frames (~5s), chunked at 25 frames (500ms) with 5 frames (100ms) lookahead.

## Module Structure

```
conformer-python/src/conformer_python/muaalem_streaming_kv_cache/
├── __init__.py
├── configuration_multi_level_ctc_streaming_kv_cache.py
├── modeling_multi_level_ctc_streaming_kv_cache.py
├── multi_level_tokenizer.py          # copy from streaming_lstm
└── vocab.py                          # copy from streaming_lstm
```

## Bold Lines (non-negotiable architectural decisions)

1. **Two modes, one model** — A single config flag `use_sliding_window` controls whether attention is full (within cached context) or sliding-window. No separate model classes.

2. **KV cache as first-class citizen** — The encoder's forward signature accepts and returns `past_key_values`. This is a list of `(key, value)` tuples, one per layer. Shape convention: `(batch, num_heads, seq_len, head_dim)`. Cache eviction policy (FIFO beyond `max_context_frames`) is the developer's choice.

3. **Chunked streaming** — The model processes 25-frame chunks. The `forward` accepts a `past_key_values` argument and returns updated `present_key_values`. The user (or a wrapper loop) is responsible for slicing audio into chunks and feeding them sequentially.

4. **No positional embeddings in encoder** — The encoder operates position-free (ergodic). Position information, if needed, is injected in an adapter/projection *after* the encoder (following Moonshine v2's philosophy). This lets the encoder be fully streaming-friendly.

5. **CTC-only decoder** — Multi-level CTC heads (same pattern as offline and streaming_lstm). Logits from consecutive chunks are concatenated along the time axis for training.

6. **Causal convolution stays** — The existing `Wav2Vec2BertConvolutionModule` is already causal (left-only padding). Reused as-is.

## Configuration additions (`Wav2Vec2BertForMultilevelCTCStreamingKVCacheConfig`)

Same as offline config, plus:

- `chunk_size` (default: 25)
- `lookahead_frames` (default: 5)
- `max_context_frames` (default: 250)
- `use_sliding_window` (default: `False`)
- `sliding_window_left` (default: 25)
- `sliding_window_right` (default: 5)
- `position_embeddings_type` overridden to `None` in encoder (ergodic)

## Model architecture (at a glance)

```
Wav2Vec2BertForMultilevelCTCStreamingKVCache
├── Wav2Vec2BertModel (with KV-cache-aware encoder)
│   ├── FeatureProjection
│   └── Wav2Vec2BertEncoder (KV-cache-aware)
│       └── Wav2Vec2BertEncoderLayer × N
│           ├── FFN1
│           ├── SelfAttention (KV-cache + optional SW mask)
│           ├── ConvModule (causal, unchanged)
│           └── FFN2
├── Adapter (optional, with position embeddings)
└── Multi-level CTC heads (linear projections)
```

## Forward pass contract

```python
def forward(
    self,
    input_features: Tensor,           # (batch, chunk_frames, feat_dim)
    attention_mask: Optional[Tensor],
    past_key_values: Optional[list]   # cached states from previous chunks
) -> tuple[Tensor, list]:
    # returns (logits, updated_past_key_values)
```

Training: pass entire utterance; model internally chunks and accumulates KV cache, concatenates per-chunk logits, computes CTC loss over the full sequence with blank insertion per Issue #10's concatenation strategy.

## Development Place
The development will take place in the [conformer-rs](https://github.com/obadx/conformer-rs) repo not here.

## After Designing

After developing the architecture run it on-device for testing

## References

- Moonshine v2 (sliding-window ergodic encoder): https://ar5iv.org/abs/2602.12241
- Issue #10 (sliding window training with CTC + blank insertion): https://github.com/obadx/quran-muaalem/issues/10
- Issue #29 (on-device plan, Architecture A): https://github.com/obadx/quran-muaalem/issues/29
- Existing streaming_lstm variant: `muaalem_streaming_lstm/` (reference for module structure)
- Moonshine toolkit (KV caching for streaming): https://github.com/moonshine-ai/moonshine

**Comment by abdelkareemkobo** [comment]
<img width="1871" height="673" alt="Image" src="https://github.com/user-attachments/assets/808c549f-291e-421d-94fc-de6fa0568a8e" />
sorry for late i will continue explore some other ideas and will let you know soon and will share code also

**Comment by obadx** [comment]
ربنا يكرمك يا  عبد الكريم <3 <3 
ممكن بس ترفع الأكود عشان في تفاصيل محاج أفهمها 
و يا ريت تعمل pr
سؤال بس إيه هو ال coding agent اللي بتستخدمع مه إنهي LLM ?

**Comment by abdelkareemkobo** [comment]
عينا هنظف بس شوية حجات ورافعها وعايز اعمل شوية تجارب بس ثانية 
بستخدمopus and deepseekv4 flash  on OpenCode Agent.

**Comment by obadx** [comment]
متنساش بس تزود معاه ال sliding window attention ك optional
